### PR TITLE
[LFXV2-1832] feat: send email notifications when a document or link is uploaded to a committee

### PR DESCRIPTION
**Pull request by andrest50**
## Summary

- Emit new `committee_document.created` and `committee_link.created` domain events from the document/link writer orchestrators after a successful create
- Subscribe to both subjects and fan out best-effort notification emails to all LFID members, writers, and auditors of the committee
- Single email template covers both file documents and links (conditional URL block for links)

## Details

**Event layer (`pkg/constants/subjects.go`, `internal/domain/model/committee_message.go`)**
- Two new event subjects following the `lfx.committee-api.<resource>.<action>` convention
- `CommitteeEvent.Build` extended with `ResourceCommitteeDocument` and `ResourceCommitteeLink` cases

**Publish on create (`internal/service/document_writer.go`, `link_writer.go`)**
- After the existing indexer publish, each writer calls `publishDocumentCreatedEvent` / `publishLinkCreatedEvent` (fire-and-forget, errors logged only)

**Subscriber + fan-out (`internal/service/document_notification_handler.go`)**
- `HandleCommitteeDocumentCreated` and `HandleCommitteeLinkCreated` both decode the event and delegate to a shared `handleContentCreated`
- `collectCommitteeRecipients` aggregates members (`ListMembers`) + writers/auditors (`GetSettings`), deduped by LFID; users without an LFID are skipped
- `errgroup` with `SetLimit(5)` fans out per recipient; missing emails backfilled via `EmailsByPrincipal`; failures logged, never propagated

**Email (`internal/service/email/committee_document_notification.go`, `templates/`)**
- `RenderCommitteeDocumentNotification` with `CommitteeDocumentNotificationData`
- One HTML + one text template; link URL shown conditionally for links, omitted for file documents
- Matches existing LFX branding (logo, `#009AFF` button, Inter font, 2026 footer)

## Ticket

[LFXV2-1832](https://linuxfoundation.atlassian.net/browse/LFXV2-1832)

## Test plan

- [x] Upload a file document to a committee that has members, writers, and auditors with LFIDs — confirm one email per distinct LFID recipient carrying the file name, committee name, and `/project/groups/<uid>` deep link
- [x] Add a link to a committee — confirm email shows the link name + URL
- [x] Verify uploader also receives the email
- [x] Verify users without an LFID (`Username == ""`) are silently skipped
- [x] Verify a single LFID that appears in multiple role lists (e.g. both member and writer) receives only one email
- [x] Verify a send failure for one recipient does not prevent emails to other recipients

🤖 Generated with [Claude Code](https://claude.ai/code)

[LFXV2-1832]: https://linuxfoundation.atlassian.net/browse/LFXV2-1832?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: internal/service/email/templates/committee_document_notification.txt [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+{{.ItemName}}{{if .ItemURL}}
+{{.ItemURL}}{{end}}{{if .ItemDescription}}
+{{.ItemDescription}}{{end}}
+
+You can view the committee here:
+{{.CommitteeURL}}
+
+---
+You are receiving this email because you are a member of a committee on LFX Self-Serve.
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/templates/committee_document_notification.txt]: Fixed — updated the footer to say "member, writer, or auditor" so it accurately reflects all roles that receive this notification. See commit 35a13cf.

File: internal/service/document_notification_handler.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+
+	item := committeeContentItem{
+		committeeUID:      link.CommitteeUID,
+		itemType:          "link",
+		itemName:          link.Name,
+		itemURL:           link.URL,
+		itemDescription:   link.Description,
+		createdByUsername: link.CreatedByUsername,
+	}
+
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:114]: Fixed — added an `isSafeURL` helper that allows only `http` and `https` schemes. If the link URL has any other scheme (e.g. `javascript:`) it is omitted from the email and a warning is logged. Added unit tests for the helper covering all edge cases. See commit 35a13cf.

File: internal/service/document_notification_handler.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+}
+
+// handleContentCreated fans out notification emails to all LFID members, writers, and auditors
+// of the committee. Best-effort: individual send failures are logged but never abort the batch.
+func (m *messageHandlerOrchestrator) handleContentCreated(ctx context.Context, item committeeContentItem) {
+	if m.committeeReader == nil || m.emailSender == nil {
+		slog.DebugContext(ctx, "committee reader or email sender not configured — skipping content notification",
+			"committee_uid", item.committeeUID)
+		return
+	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:126]: Done — added `document_notification_handler_test.go` covering: (1) dedupe across members/writers/auditors, (2) skipping `Username==""`, (3) email backfill via `EmailsByPrincipal`, (4) send failures not aborting the batch, (5) unsafe URL filtering. See commit 35a13cf.

File: internal/service/document_notification_handler.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+		}
+		if _, ok := seen[username]; ok {
+			return
+		}
+		seen[username] = struct{}{}
+		recipients = append(recipients, notificationRecipient{
+			username: username,
+			email:    email,
+			name:     name,
+		})
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:254]: Fixed — changed `seen` from `map[string]struct{}` to `map[string]int` (index into the recipients slice). When a username is seen again, the existing entry is enriched with a non-empty email or name from the later record rather than being dropped. Exactly as proposed. See commit 35a13cf.

File: Makefile [hunk 5]
```diff
@@ -11,7 +11,7 @@ GIT_COMMIT := $(shell git rev-parse HEAD)
 DOCKER_REGISTRY := ghcr.io/linuxfoundation
 DOCKER_IMAGE := $(DOCKER_REGISTRY)/$(APP_NAME)/committee-api
 DOCKER_CLI_IMAGE := $(DOCKER_REGISTRY)/$(APP_NAME)/committee-cli
-DOCKER_TAG ?= $(VERSION)
+DOCKER_TAG ?= latest
```
[the remark this answers is not in the document]
Comment (andrest50) [reply Makefile:14]: Good catch on the traceability concern. This change is inherited from the base branch (`feature/email-invite-feature-flags`) and was not introduced by this PR — it will be addressed separately in that branch before it merges to main. No change needed here.

[the remark this answers is not in the document]
Comment (andrest50) [reply Makefile:14]: This `DOCKER_TAG` change is inherited from the base branch (`feature/email-invite-feature-flags`) and was not introduced by this PR's commits. It will be addressed in that branch separately.

File: internal/service/document_notification_handler.go [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	recipients := m.collectCommitteeRecipients(ctx, item.committeeUID)
+	if len(recipients) == 0 {
+		slog.DebugContext(ctx, "no LFID recipients for content notification", "committee_uid", item.committeeUID)
+		return
+	}
+
+	uploaderName := m.resolveDisplayNameWithTimeout(ctx, item.createdByUsername)
+	committeeURL := buildCommitteeURL(m.lfxSelfServeBaseURL, item.committeeUID)
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go]: Fixed — `uploaderName` is now only resolved when `createdByUsername` is non-empty. When there is no principal, `UploaderName` stays empty and the template renders the generic "A new document/link was added to..." message. See commit 35a13cf.

File: internal/service/document_notification_handler.go [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+	}
+
+	members, err := m.committeeReader.ListMembers(ctx, committeeUID)
+	if err != nil {
+		slog.WarnContext(ctx, "failed to list members for content notification",
+			"error", err, "committee_uid", committeeUID)
+	} else {
+		for _, m := range members {
+			add(m.Username, m.Email, strings.TrimSpace(m.FirstName+" "+m.LastName))
+		}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go]: Fixed — renamed the loop variable from `m` to `member` in `collectCommitteeRecipients`. See commit 35a13cf.

File: internal/service/email/templates/committee_document_notification.txt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+Name: {{.DocumentName}}
+{{if eq .DocumentType "file"}}File: {{.FileName}}
+{{end}}{{if and (eq .DocumentType "link") .URL}}URL: {{.URL}}
+{{end}}{{if .FolderName}}Folder: {{.FolderName}}
+{{end}}
+You can view the committee and its documents here:
+{{.CommitteeURL}}
+
+---
+You are receiving this email because you are a member of a committee on LFX Self-Serve.
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/templates/committee_document_notification.txt]: Fixed — updated the footer to say "member, writer, or auditor" and also aligned "documents and links" to "content" for accuracy. See commit 35a13cf.

File: internal/service/email/templates/committee_document_notification.html [hunk 9]
```diff
@@ -0,0 +1,64 @@
+<!-- Copyright The Linux Foundation and each contributor to LFX. -->
+<!-- SPDX-License-Identifier: MIT -->
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <meta name="viewport" content="width=device-width, initial-scale=1.0">
+    <meta name="description" content="LFX Self-Serve committee document notification">
+    <meta name="keywords" content="LFX, committee, document, notification">
+    <title>New document added to {{.CommitteeName}}</title>
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/templates/committee_document_notification.html]: Fixed — the `<title>` now uses `{{if eq .DocumentType "link"}}link{{else}}document{{end}}` to match the body copy. See commit 35a13cf.

File: internal/service/email/templates/committee_document_notification.html [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+    <p class="greeting">Hi {{.RecipientName}},</p>
+
+    {{if .UploaderName}}<p>{{.UploaderName}} has added a new {{if eq .DocumentType "link"}}link{{else}}document{{end}} to the <strong>{{.CommitteeName}}</strong> committee in LFX Self-Serve.</p>{{else}}<p>A new {{if eq .DocumentType "link"}}link{{else}}document{{end}} has been added to the <strong>{{.CommitteeName}}</strong> committee in LFX Self-Serve.</p>{{end}}
+
+    <p class="detail-item"><span class="detail-label">Name:</span> {{.DocumentName}}</p>
+    {{if eq .DocumentType "file"}}<p class="detail-item"><span class="detail-label">File:</span> {{.FileName}}</p>{{end}}
+    {{if and (eq .DocumentType "link") .URL}}<p class="detail-item"><span class="detail-label">URL:</span> <a href="{{.URL}}" class="detail-link">{{.URL}}</a></p>{{end}}
+    {{if .FolderName}}<p class="detail-item"><span class="detail-label">Folder:</span> {{.FolderName}}</p>{{end}}
+
+    <p class="details-footer">You can view the committee and its documents by visiting the link below.</p>
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/templates/committee_document_notification.html]: Fixed — changed the pre-button sentence to "You can view the committee and its content by visiting the link below." See commit 35a13cf.

File: internal/service/email/committee_document_notification.go [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+	DocumentType  string // "file" | "link"
+	DocumentName  string // user-given display name
+	FileName      string // actual file name, set for DocumentType "file"
+	URL           string // set for DocumentType "link"
+	FolderName    string // optional — set when the item lives inside a folder
+}
+
+// RenderCommitteeDocumentNotification renders the subject, HTML body, and plain-text body for an
+// email notifying committee members/writers/auditors that a new document or link was added.
+func RenderCommitteeDocumentNotification(data CommitteeDocumentNotificationData) (subject, html, text string, err error) {
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/committee_document_notification.go:43]: Done — added `committee_document_notification_test.go` covering: file document, link (URL row shown), folder row, no-uploader generic subject/body, conditional HTML title, and footer role wording. See commit 35a13cf.

File: internal/service/document_notification_handler.go [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+	createdByUsername string // LFID of the uploader/creator
+}
+
+// HandleCommitteeDocumentCreated handles committee_document.created events and notifies all
+// LFID members, writers, and auditors of the committee. Best-effort: errors are logged, not returned.
+func (m *messageHandlerOrchestrator) HandleCommitteeDocumentCreated(ctx context.Context, msg port.TransportMessenger) ([]byte, error) {
+	var event model.CommitteeEvent
+	if err := json.Unmarshal(msg.Data(), &event); err != nil {
+		slog.WarnContext(ctx, "failed to unmarshal committee_document.created event", "error", err)
+		return nil, nil
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:38]: Done — added `document_notification_handler_test.go` with full handler coverage. See commit 35a13cf.

**Comment by andrest50** [comment]
## Review Feedback Addressed

Commit: 35a13cfce032c1c2f4020d2c4c5c450569257f0e

### Changes Made

- **`document_notification_handler.go`**: Sanitize link URL — only `http`/`https` schemes are included in emails; unsafe schemes (e.g. `javascript:`) are omitted and logged (per copilot-pull-request-reviewer)
- **`document_notification_handler.go`**: Only resolve uploader display name when `createdByUsername` is non-empty — fixes misleading \"A committee administrator has added...\" copy when there is no principal (per copilot-pull-request-reviewer)
- **`document_notification_handler.go`**: Rename loop variable `m` → `member` in `collectCommitteeRecipients` to avoid shadowing the orchestrator receiver (per copilot-pull-request-reviewer)
- **`document_notification_handler.go`**: Improve dedup — `seen` map now stores the index into the recipients slice so later records can enrich a missing email or name rather than being dropped (per coderabbitai[bot])
- **`templates/committee_document_notification.html`**: Make `<title>` conditional on `DocumentType` so link notifications show \"New link added to...\" (per copilot-pull-request-reviewer)
- **`templates/committee_document_notification.html`**: Change \"documents\" → \"content\" in the pre-button sentence for accuracy with links (per copilot-pull-request-reviewer)
- **`templates/committee_document_notification.txt`**: Update footer to say \"member, writer, or auditor\" instead of just \"member\" (per copilot-pull-request-reviewer, coderabbitai[bot])

### Tests Added

- **`email/committee_document_notification_test.go`**: Render tests for file document, link (URL row), folder row, no-uploader generic copy, conditional HTML title, and footer role wording (per copilot-pull-request-reviewer)
- **`service/document_notification_handler_test.go`**: Handler tests covering dedupe across roles, no-LFID skip, email backfill via `EmailsByPrincipal`, send-failure best-effort, unsafe URL filtering, and `isSafeURL` unit tests (per copilot-pull-request-reviewer, coderabbitai[bot])

### No Change Needed

- **`Makefile:14`** (`DOCKER_TAG ?= latest`): This change comes from the base branch (`feature/email-invite-feature-flags`), not from this PR's commits. It will be addressed in that branch before it merges to main.

### Threads Resolved

11 of 13 unresolved threads addressed in this iteration.

### Still Open

- **`Makefile:14`** (×2, coderabbitai[bot] + copilot-pull-request-reviewer): Left open — inherited from base branch, outside scope of this PR.

File: internal/service/email/templates/committee_document_notification.txt [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+Hi {{.RecipientName}},
+
+{{if .UploaderName}}{{.UploaderName}} has added a new {{if eq .DocumentType "link"}}link{{else}}document{{end}} to the {{.CommitteeName}} committee in LFX Self-Serve.{{else}}A new {{if eq .DocumentType "link"}}link{{else}}document{{end}} has been added to the {{.CommitteeName}} committee in LFX Self-Serve.{{end}}
+
+Name: {{.DocumentName}}
+{{if eq .DocumentType "file"}}File: {{.FileName}}
+{{end}}{{if and (eq .DocumentType "link") .URL}}URL: {{.URL}}
+{{end}}{{if .FolderName}}Folder: {{.FolderName}}
+{{end}}
+You can view the committee and its documents here:
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/email/templates/committee_document_notification.txt]: Fixed — changed "its documents" to "its content" in the plain-text template to match the HTML template. Good catch — we missed this one when fixing the HTML. See commit 64baf27.

**Comment by andrest50** [comment]
## Review Feedback Addressed

Commit: 64baf277d14df11b39a0f7eb818d71ad5a2eb981

### Changes Made

- **`templates/committee_document_notification.txt:12`**: Changed "its documents" → "its content" in the plain-text pre-link sentence, matching the HTML template fix from the previous iteration (per copilot-pull-request-reviewer)

### Threads Resolved

1 of 1 new unresolved thread addressed in this iteration.

File: internal/service/document_notification_handler_test.go [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+			lfxSelfServeBaseURL: "https://app.dev.lfx.dev",
+		}
+		msg := newMockTransportMessenger(constants.CommitteeDocumentCreatedSubject, buildDocumentCreatedPayload(t, doc))
+		_, _ = h.HandleCommitteeDocumentCreated(context.Background(), msg)
+
+		require.NotEmpty(t, sender.calls)
+		assert.Contains(t, sender.calls[0].Subject, "TSC Committee")
+		assert.Contains(t, sender.calls[0].HTML, "Q1 Report")
+		assert.Contains(t, sender.calls[0].HTML, "https://app.dev.lfx.dev/project/groups/committee-1")
+	})
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler_test.go:244]: Fixed — replaced `_, _ =` with `resp, err :=` and added `assert.NoError(t, err)` + `assert.Nil(t, resp)` before the content assertions. See commit 056b82c.

**Comment by andrest50** [comment]
## Review Feedback Addressed

Commit: 056b82c4470913a857c113e12c95e95974584290

### Changes Made

- **`document_notification_handler_test.go:237`**: Replaced `_, _ =` with `resp, err :=` and added `assert.NoError`/`assert.Nil` assertions in the content-assertions subtest (per coderabbitai[bot])

### Threads Resolved

1 of 1 new unresolved thread addressed in this iteration.

File: internal/service/document_notification_handler.go [hunk 15]
```diff
@@ -0,0 +1,304 @@
...
+	if err != nil {
+		slog.WarnContext(ctx, "committee_document.created event has unexpected data shape", "error", err)
+		return nil, nil
+	}
+
+	var doc model.CommitteeDocument
+	if err := json.Unmarshal(raw, &doc); err != nil {
+		slog.WarnContext(ctx, "cannot decode CommitteeDocument from event data", "error", err)
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:51]: Fixed — added an early-return guard in `HandleCommitteeDocumentCreated` after decoding: if `CommitteeUID` is empty the event is discarded with a `WarnContext` log. Matches the pattern from other handlers in the service. See commit 8e75831.

File: internal/service/document_notification_handler.go [hunk 16]
```diff
@@ -0,0 +1,304 @@
...
+	if err != nil {
+		slog.WarnContext(ctx, "committee_link.created event has unexpected data shape", "error", err)
+		return nil, nil
+	}
+
+	var link model.CommitteeLink
+	if err := json.Unmarshal(raw, &link); err != nil {
+		slog.WarnContext(ctx, "cannot decode CommitteeLink from event data", "error", err)
+		return nil, nil
+	}
```
[the remark this answers is not in the document]
Comment (andrest50) [reply internal/service/document_notification_handler.go:90]: Fixed — added an early-return guard in `HandleCommitteeLinkCreated` after decoding: if `CommitteeUID` is empty the event is discarded with a `WarnContext` log. Matches the pattern from other handlers in the service. See commit 8e75831.

**Comment by andrest50** [comment]
## Review Feedback Addressed

Commit: 8e75831b7fd61ae61299824c04c99c794f13e2b4

### Changes Made

- **`document_notification_handler.go:51`**: Added early-return guard in `HandleCommitteeDocumentCreated` — discards events with empty `CommitteeUID` with a `WarnContext` log before any downstream work (per copilot-pull-request-reviewer)
- **`document_notification_handler.go:85`**: Same guard added in `HandleCommitteeLinkCreated` (per copilot-pull-request-reviewer)

### Threads Resolved

2 of 2 new unresolved threads addressed in this iteration.

File: internal/service/document_notification_handler.go [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+
+	committee, _, err := m.committeeReader.GetBase(ctx, item.committeeUID)
+	if err != nil {
+		slog.WarnContext(ctx, "failed to load committee for content notification",
+			"error", err, "committee_uid", item.committeeUID)
+		return
+	}
+
+	recipients := m.collectCommitteeRecipients(ctx, item.committeeUID)
```
Comment (dealako) [inline internal/service/document_notification_handler.go:135]: [minor] **Uploader receives a notification about their own upload — confirm if intentional**

`collectCommitteeRecipients` adds all members + writers + auditors without excluding the actor (`item.createdByUsername` / `UploadedByUsername`). The person who just uploaded the document or link will receive a notification email about their own action.

The PR description says fan out to "**all**" roles, so this may be by design. If it is intentional, a brief comment here would clarify the intent. If unintended, you can skip the actor:

```go
add := func(username, email, name string) {
    if username ==  {
        return
    }
    // Skip the actor — they know what they uploaded.
    if username == item.createdByUsername {
        return
    }
    // ... rest of dedupe logic
}
```

Not raised by prior bot reviewers.

**Review by dealako — DISMISSED** [review]
Hi Andres! 👋

This is a clean, well-considered implementation of the document/link notification pipeline. The event-driven architecture fits naturally into the existing pattern, the fan-out logic is solid, and the test coverage that landed in the review rounds is thorough. A few observations below — two minor items neither of which blocks merge.

---

## Revision Tracking

All 9 prior round issues from Copilot and CodeRabbit have been fully resolved:

| Prior issue | Round | Resolved in |
|---|---|---|
| Unsafe URL scheme (`javascript:`) in emailed links | Copilot R1 | `35a13cf` |
| Missing handler + email render tests | Copilot R1 | `35a13cf` |
| Dedupe data loss (first-seen wins over richer email/name) | CodeRabbit R1 | `35a13cf` |
| Template wording ("documents" → "content"), conditional `<title>`, footer roles | Copilot R1 + R2 | `35a13cf`, `64baf27` |
| Empty uploader producing misleading "A committee administrator uploaded" copy | Copilot R1 | `35a13cf` |
| Loop variable `m` shadowing receiver | Copilot R1 | `35a13cf` |
| Test discards `resp`/`err` return values | CodeRabbit R3 | `056b82c` |
| Missing `committee_uid` guard on both handlers | Copilot R4 | `8e75831` |

---

## AI Bot Reconciliation

I independently reviewed the code and agree with all bot findings and their resolutions — no re-litigation needed. Specifically:

- **`isSafeURL`** (`35a13cf`) — correct fix; `url.Parse` + scheme allowlist is idiomatic Go and the scheme comparison is properly case-folded via `strings.ToLower`. ✅
- **Dedupe enrichment** (`35a13cf`) — `map[string]int` (index into slice) is an elegant approach; confirmed the `add` closure correctly enriches existing entries rather than dropping them. ✅
- **`committee_uid` empty guard** (`8e75831`) — matches the exact pattern used by `HandleCommitteeUpdated`, `HandleCommitteeMailingListChanged`, etc. ✅
- **Makefile `DOCKER_TAG → latest`** — confirmed this diff is inherited from the `feature/email-invite-feature-flags` base branch; the PR's own commits do not touch Makefile behaviour. No action needed here. ✅

---

## Security Review

Ran `/lfx-security-engineer` across all 20 changed files. **No blockers found.**

**Passed checks:**
- ✅ No hardcoded secrets, tokens, or credentials
- ✅ HTML body uses `html/template` (contextual auto-escaping for `DocumentName`, `FileName`, `URL`, `CommitteeURL`)
- ✅ Email subject sanitised via `sanitizeHeader()` on uploader name and committee name
- ✅ `isSafeURL()` provides defence-in-depth on top of `html/template`'s href sanitisation
- ✅ Username logged only through `redaction.Redact()` — no raw PII in log output
- ✅ `CommitteeURL` built from a config-supplied base URL + a UUID validated at upload time — no open redirect risk
- ✅ No SQL/command injection patterns; no template-injection bypass (`safeHTML`, `[innerHTML]`)
- ✅ `text/template` used for plain-text body only — no injection risk in that context

---

## New Findings (2 items, neither blocking)

See inline comments for detail.

---

## Summary

**Decision: ✅ Approved with minor comments**

0 blocking issues · 1 minor (self-notification) · 1 nit (backfill cost). The implementation is correct, consistent with sibling handlers, and the test suite covers the important paths well. Great work on responding to the review rounds so thoroughly, Andres!

File: internal/service/document_notification_handler.go [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+		g.Go(func() error {
+			email := recipient.email
+			if email == "" && m.userReader != nil && recipient.username != "" {
+				lookupCtx, cancel := context.WithTimeout(gctx, committeeNotificationTimeout)
+				emails, lookupErr := m.userReader.EmailsByPrincipal(lookupCtx, recipient.username)
+				cancel()
+				if lookupErr == nil && emails != nil && emails.PrimaryEmail != "" {
+					email = emails.PrimaryEmail
+				}
+			}
```
Comment (dealako) [inline internal/service/document_notification_handler.go:163]: [nit] **Per-recipient email backfill serialises in waves for large committees**

When a recipient has no stored email, this block fires a `EmailsByPrincipal` RPC call for each such user (bounded by `SetLimit(5)`, so waves of 5 concurrent lookups, each up to 5 s). For a typical committee this is fine. For a very large committee (e.g. 200+ members, many without stored email) the fan-out could be slow or chatty.

Worth noting as a future optimisation: batch-resolving emails for all recipients missing an address before the goroutine fan-out, if `EmailsByPrincipal` ever supports a bulk variant. No action needed now — the async, best-effort nature of notifications makes this a latency concern rather than a correctness one.

File: internal/service/document_notification_handler.go (hunk 17 above)
Comment (andrest50) [reply internal/service/document_notification_handler.go:135]: This is intentional by design — the decision was made to notify all roles uniformly (members, writers, auditors) regardless of who uploaded, so the document appears in everyone's inbox consistently. I've added a comment to `collectCommitteeRecipients` to make this explicit. See commit 8534f76.

File: internal/service/document_notification_handler.go (hunk 18 above)
Comment (andrest50) [reply internal/service/document_notification_handler.go:163]: Agreed — for typical committee sizes this is fine, and the async best-effort nature of notifications means latency here isn't a correctness concern. Noted as a future optimization: if `EmailsByPrincipal` ever supports a bulk variant, batch-resolving emails before the goroutine fan-out would be the right approach. No change now.

**Comment by andrest50** [comment]
## Review Feedback Addressed — @dealako

Commit: 8534f76a80b7bf66158e79817d57968569a0bd30

### Changes Made

- **`document_notification_handler.go`**: Added a comment to `collectCommitteeRecipients` clarifying that the uploader is intentionally included in the recipient list — the notification is sent uniformly to all roles (members, writers, auditors) regardless of who uploaded the item (per @dealako)

### No Change Needed

- **`document_notification_handler.go:163`**: Per-recipient email backfill in goroutine waves is acknowledged as a known future optimization. For current committee sizes it's not a concern, and the async best-effort nature means latency here isn't a correctness issue. Will revisit if `EmailsByPrincipal` adds a bulk variant (flagged by @dealako — no action needed per reviewer)

### Threads Resolved

2 of 2 threads from @dealako addressed in this iteration.

**Comment by dealako** [comment]
Hi Andres! 👋 Thanks for turning this around quickly — only two small commits needed to close out the review round, which says a lot about the quality of the original implementation.

---

### 👏 Nice work

The documentation instinct here is exactly right. Rather than filtering the uploader out (which would have been a one-liner but buried the reasoning), you wrote down *why* the behavior is what it is. Future maintainers will thank you when they stumble on this and wonder the same thing I did.

---

### ✅ Resolved

**[minor] Uploader self-notification** — `internal/service/document_notification_handler.go:229`
Resolved in commit `8534f76a`. The added comment clearly captures the intent: uniform fan-out to all roles regardless of who performed the upload, so every committee member sees the document in their inbox consistently. That's a completely valid product decision and the comment makes it explicit. ✅

**[nit] Per-recipient email backfill in waves** — `document_notification_handler.go:163`
Acknowledged and accepted as-is in your reply comment. Appropriate disposition — this is a future optimization, not a correctness issue, and current committee sizes make it a non-problem. ✅

---

### ⚠️ Partially addressed

_None._

---

### ❌ Still open

_None._

---

### 🔍 New observations

**`main` merge (`7718d7f4`)** — touched only `.github/workflows/ko-build-tag.yaml`, `.github/workflows/weekly-brief-eval-live.yml`, `README.md`, and `evals/weekly-brief/README.md`. No feature-code changes. The CI workflow changes add a weekly-brief live-eval release gate — nothing that affects this feature or introduces new security concerns.

**Security re-scan** — ran `/lfx-security-engineer` on the full delta since the prior review. No new findings. Exit code 0. ✅

---

### Overall decision: ✅ Approved

Both prior findings are fully resolved. CI is green (10/10 checks pass). No regressions, no new issues, no merge conflicts. This is ready to merge.

**Review by dealako — APPROVED** [review]
Re-approving after verifying the two follow-up commits. Both prior findings resolved — the clarifying comment on uploader intent is exactly what was asked for, and the nit disposition is appropriate. CI green, no new concerns. Ship it.
